caffe2/opt/optimize_ideep.cc

```diff
@@ -1,51 +1,377 @@
 #include "caffe2/opt/optimize_ideep.h"
 #include "caffe2/opt/converter.h"
 #include "caffe2/opt/fusion.h"
-#include "caffe2/utils/proto_utils.h"
+
+#ifdef CAFFE2_USE_IDEEP
+#include "caffe2/ideep/ideep_utils.h"
+#endif
 
 namespace caffe2 {
 namespace opt {
 
 using namespace nom;
 
-void OptimizeForIdeep(repr::NNModule* nn) {
-  // Conv+Relu fusion
-  auto should_fuse = [](const repr::Conv& conv) {
-    const auto annotation = conv.getAnnotation();
-    if (!annotation || !isa<Caffe2Annotation>(annotation)) {
-      return false;
+#ifndef CAFFE2_USE_IDEEP
+void OptimizeForIdeep(repr::NNModule* nn, caffe2::Workspace* ws, bool training_mode) {
+  LOG(WARNING) << "Only support optimizations for IDEEP";
+}
+
+#else
+USE_IDEEP_DEF_ALIASES();
+
+Blob* getBlob(repr::NNGraph::NodeRef node, caffe2::Workspace* ws) {
+  auto tensor = repr::nn::get<repr::Tensor>(node);
+  assert(ws->HasBlob(tensor->getName()) && "Blob not in workspace");
+  return ws->GetBlob(tensor->getName());
+}
+
+template<class T>
+T* getTensor(Blob* blob) {
+  assert(blob && "Blob is invalid");
+  if (blob && blob->template IsType<T>()) {
+    return blob->template GetMutable<T>();
+  }
+  return nullptr;
+}
+
+const caffe2::OperatorDef* getOpDef(const repr::NeuralNetOperator& nnOp) {
+  auto annotation = nnOp.getAnnotation();
+  if (!annotation || !isa<Caffe2Annotation>(annotation)) {
+    return nullptr;
+  }
+
+  return dyn_cast<Caffe2Annotation>(annotation)->getOperatorDef();
```

Comment:
It no longer returns pointer. Please rebase against recent master.

Reply:
fixed